tests/keras/layers/convolutional_test.py

```diff
@@ -21,145 +19,136 @@
 @keras_test
 @pytest.mark.skipif((K.backend() == 'cntk'),
                     reason="cntk does not support dilated conv")
-def test_causal_dilated_conv():
-    # Causal:
-    layer_test(convolutional.Conv1D,
-               input_data=np.reshape(np.arange(4, dtype='float32'), (1, 4, 1)),
-               kwargs={
-                   'filters': 1,
-                   'kernel_size': 2,
-                   'dilation_rate': 1,
-                   'padding': 'causal',
-                   'kernel_initializer': 'ones',
-                   'use_bias': False,
-               },
-               expected_output=[[[0], [1], [3], [5]]]
-               )
-
-    # Non-causal:
-    layer_test(convolutional.Conv1D,
-               input_data=np.reshape(np.arange(4, dtype='float32'), (1, 4, 1)),
-               kwargs={
-                   'filters': 1,
-                   'kernel_size': 2,
-                   'dilation_rate': 1,
-                   'padding': 'valid',
-                   'kernel_initializer': 'ones',
-                   'use_bias': False,
-               },
-               expected_output=[[[1], [3], [5]]]
-               )
-
-    # Causal dilated with larger kernel size:
-    layer_test(convolutional.Conv1D,
-               input_data=np.reshape(np.arange(10, dtype='float32'), (1, 10, 1)),
-               kwargs={
-                   'filters': 1,
-                   'kernel_size': 3,
-                   'dilation_rate': 2,
-                   'padding': 'causal',
-                   'kernel_initializer': 'ones',
-                   'use_bias': False,
-               },
-               expected_output=np.float32(
-                   [[[0], [1], [2], [4], [6], [9], [12], [15], [18], [21]]])
-               )
+@pytest.mark.parametrize(
+    'input_length,kernel_size,dilation_rate,padding,expected_output',
+    [
+        (4, 2, 1, 'causal', [[[0], [1], [3], [5]]]),  # Causal
+        (4, 2, 1, 'valid', [[[1], [3], [5]]]),  # Non-causal
+        # Causal dilated with larger kernel size
+        (10, 3, 2, 'causal', np.float32([[[0], [1], [2], [4], [6],
+                                          [9], [12], [15], [18], [21]]])),
+    ]
+)
+def test_causal_dilated_conv(
+        input_length, kernel_size, dilation_rate, padding, expected_output):
```

Comment:
I believe this parameterization pattern is not ideal for readability. It is not easy to assess at a glance what the tests cases are, and it is not easy to add new ones.

I suggest the following pattern for better readability:

```python
@pytest.mark.parametrize('layer_kwargs,input_length,output_length',
   [({'filters': 3, ...}, 5, 3),
     ...])
def test_causal_dilated_conv(layer_kwargs, input_shape, output_shape):
    input_data = ...
    expected_output = ...
    layer_test(convolutional.Conv1D, input_data=input_data, kwargs=layer_kwargs, expected_output=expected_output)
```

That way the different layer configurations are enumerated as dictionaries: you immediately know the value of each named argument.

What do you think?

Reply:
Yes, indeed. You approach is more readable. I will fix it later.